### PR TITLE
Feature/issues/344

### PR DESCRIPTION
**Pull request by Mizzick**
#344

### CODE REVIEW

File: Extension/browser/webkit/lib/utils/csp-utils.js [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+
+        /*
+         https://bugs.chromium.org/p/chromium/issues/detail?id=513860
+         Bad Chromium bug: web pages can work around CSP directives by
+         creating data:- or blob:-based URI. So if we must restrict using CSP,
+         we have no choice but to also prevent the creation of nested browsing
+         contexts based on data:- or blob:-based URIs.
+         https://github.com/AdguardTeam/AdguardBrowserExtension/issues/344
+         */
+        if (Prefs.platform === "chromium" && blockWebSockets) {
```
Comment (ameshkov) [inline Extension/browser/webkit/lib/utils/csp-utils.js]: What's the point in this check? This code is already used in Chromium only.

File: Extension/browser/webkit/lib/utils/csp-utils.js [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+    var reReportDirective = /report-(?:to|uri)[^;]*;?\s*/;
+    var reEmptyDirective = /^([a-z-]+)\s*;/;
+
+    /**
+     * Fills csp header with websocket blocking directives.
+     *
+     * @param headers headers collection
+     * @param blockWebSockets block websocket flag
+     */
+    var writeCspHeader = function (headers, blockWebSockets) {
```
Comment (ameshkov) [inline Extension/browser/webkit/lib/utils/csp-utils.js]: 1. Remove "blockWebSockets" parameter
2. Rename method to "blockWebSockets"

File: Extension/browser/webkit/webrequest.js [hunk 3]
```diff
@@ -92,11 +101,35 @@ function onHeadersReceived(requestDetails) {
     webRequestService.processRequestResponse(tab, requestUrl, referrerUrl, requestType, responseHeaders);
 
     if (requestType == RequestTypes.DOCUMENT) {
-        //safebrowsing check
+        // Safebrowsing check
         filterSafebrowsing(tab, requestUrl);
+
+        /*
+         Websocket check,
+         ff 'ws://' request is blocked for not existing domain - it's blocked for all domains.
```
Comment (ameshkov) [inline Extension/browser/webkit/webrequest.js]: ff --> if

**Comment by Mizzick** [comment]
Ok, done

File: Extension/browser/webkit/webrequest.js [hunk 4]
```diff
@@ -92,11 +101,34 @@ function onHeadersReceived(requestDetails) {
...
+         If 'ws://' request is blocked for not existing domain - it's blocked for all domains.
+         Then we gonna limit frame sources to http to block src:'data/text' etc.
+         More details in the issue:
+         https://github.com/AdguardTeam/AdguardBrowserExtension/issues/344
+
+         WS connections are detected as "other"  by ABP
+         EasyList already contains some rules for WS connections with $other modifier
+         */
+        var websocketCheckUrl = "ws://adguardwebsocket.check/";
+        if (webRequestService.checkWebSocketRequest(tab, websocketCheckUrl, referrerUrl, RequestTypes.OTHER)) {
```
Comment (ameshkov) [inline Extension/browser/webkit/webrequest.js]: I suppose we should remove `requestType` parameter from this method.

WebSocket connections are of type `OTHER`.

Comment (Mizzick) [inline Extension/browser/webkit/webrequest.js]: ok

**Pull request merged**